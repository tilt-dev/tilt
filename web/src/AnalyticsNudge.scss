@import "constants";

.AnalyticsNudge {
  width: 100%;
  min-height: $analyticsNudge-height;
  background-color: $color-purple;
  opacity: 0.98;
  color: $color-gray-darkest;
  padding: $spacing-unit / 2;
  box-sizing: border-box;
  position: fixed;
  z-index: $z-analyticsNudge;
  border: 1px solid $color-gray-light;
  box-shadow: 0px 3px 10px 0px rgba($color-gray-dark, $translucent-ish);

  // by default, it's hidden
  display: flex;
  justify-content: space-between;
  align-items: center;
  top: -100%;
  transition: top $animation-timing ease;

  pre {
    display: inline;
    font-weight: bold;
  }

  &.is-visible {
    top: 0;
  }
}

.AnalyticsNudge-button {
  align-content: center;
  background-color: $color-gray-dark;
  color: $color-gray-lightest;
  @include button-text;
  border: 1px solid $color-gray-light;
  padding: $spacing-unit/3;
  margin-left: $spacing-unit/2;
  min-width: $spacing-unit * 3;
}

.AnalyticsNudge-button:hover {
  background-color: $color-gray-light;
  cursor: pointer;
}
.opt-in {
  color: $color-green;
}

.AnalyticsNudge-button + .AnalyticsNudge-button {
  margin-left: $spacing-unit/2;
}
