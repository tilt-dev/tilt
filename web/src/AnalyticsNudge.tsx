import React, { Component } from "react"
import "./AnalyticsNudge.scss"

type AnalyticsNudgeProps = {
  needsNudge: boolean
}

type AnalyticsNudgeState = {
  requestMade: boolean
  responseCode: number
  dismissed: boolean
}

// constructor for defaults?

class AnalyticsNudge extends Component<
  AnalyticsNudgeProps,
  AnalyticsNudgeState
> {
  shouldShow(): boolean {
    return (
      this.props.needsNudge ||
      (this.state && this.state.requestMade && !this.state.dismissed)
    )
  }
  analyticsOpt(optIn: boolean) {
    let url = `//${window.location.host}/api/analytics_opt`

    let payload = { opt: optIn ? "opt-in" : "opt-out" }

    this.setState({ requestMade: true })

    fetch(url, {
      method: "post",
      body: JSON.stringify(payload),
    }).then((response: Response) => {
      this.setState({
        responseCode: response.status,
      })

      // after 3s, dismiss the nudge
      setTimeout(() => {
        this.setState({ dismissed: true })
      }, 3000)
    })
  }

  messageElem(): JSX.Element {
    if (this.state && this.state.responseCode) {
      if (this.state.responseCode == 200) {
        // Successfully called opt endpt.
        return <span>Cool, got it! üëç</span>
      } else {
        return (
          // error calling the opt endpt.
          <span>
            Oh no, something went wrong!{" "}
            <a href="https://tilt.dev/contact" target="_blank">
              Get in touch
            </a>
            .
          </span>
        )
      }
    }

    if (this.state && this.state.requestMade) {
      // request in progress
      return <span>Okay, we'll inform the robots...</span>
    }
    return (
      <div>
        <span>
          Congrats on your first Tilt resource üéâ Opt into analytics? (Read more{" "}
          <a href="https://github.com/windmilleng/tilt#privacy" target="_blank">
            here
          </a>
          .)&nbsp;
        </span>
        <span className="AnalyticsNudge-buttons">
          <button onClick={() => this.analyticsOpt(true)}>Yes!</button>
          <button onClick={() => this.analyticsOpt(false)}>Nope!</button>
        </span>
      </div>
    )
  }
  render() {
    let classes = ["AnalyticsNudge"]
    if (this.shouldShow()) {
      // or if already visible...
      classes.push("is-visible")
    }

    return <div className={classes.join(" ")}>{this.messageElem()}</div>
  }
}

export default AnalyticsNudge
