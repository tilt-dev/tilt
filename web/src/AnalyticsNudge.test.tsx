import React from "react"
import AnalyticsNudge from "./AnalyticsNudge"
import { shallow } from "enzyme"

it("hides nudge if !needsNudge and no state", () => {
  const component = shallow(<AnalyticsNudge needsNudge={false} />)

  expect(component).toMatchSnapshot()
})

it("shows nudge if needsNudge", () => {
  const component = shallow(<AnalyticsNudge needsNudge={true} />)

  expect(component).toMatchSnapshot()
})

it("shows request-in-progress message", () => {
  const component = shallow(<AnalyticsNudge needsNudge={false} />)

  component.setState({ requestMade: true })

  expect(component).toMatchSnapshot()
})

it("shows success message", () => {
  const component = shallow(<AnalyticsNudge needsNudge={false} />)

  component.setState({ requestMade: true, responseCode: 200 })

  expect(component).toMatchSnapshot()
})

it("shows failure message", () => {
  const component = shallow(<AnalyticsNudge needsNudge={false} />)

  component.setState({ requestMade: true, responseCode: 418 })

  expect(component).toMatchSnapshot()
})

it("hidden if dismissed", () => {
  const component = shallow(<AnalyticsNudge needsNudge={false} />)

  component.setState({ requestMade: true, responseCode: 200, dismissed: true })

  expect(component).toMatchSnapshot()
})
