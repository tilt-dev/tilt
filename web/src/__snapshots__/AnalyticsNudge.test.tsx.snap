// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hidden if dismissed 1`] = `
<div
  className="AnalyticsNudge"
>
  <span>
    Cool, got it! 👍
  </span>
</div>
`;

exports[`hides nudge if !needsNudge and no state 1`] = `
<div
  className="AnalyticsNudge"
>
  <div>
    <span>
      Congrats on your first Tilt resource 🎉 Opt into analytics? (Read more
       
      <a
        href="https://github.com/windmilleng/tilt#privacy"
        target="_blank"
      >
        here
      </a>
      .) 
    </span>
    <span
      className="AnalyticsNudge-buttons"
    >
      <button
        onClick={[Function]}
      >
        Yes!
      </button>
      <button
        onClick={[Function]}
      >
        Nope!
      </button>
    </span>
  </div>
</div>
`;

exports[`shows failure message 1`] = `
<div
  className="AnalyticsNudge is-visible"
>
  <span>
    Oh no, something went wrong!
     
    <a
      href="https://tilt.dev/contact"
      target="_blank"
    >
      Get in touch
    </a>
    .
  </span>
</div>
`;

exports[`shows nudge if needsNudge 1`] = `
<div
  className="AnalyticsNudge is-visible"
>
  <div>
    <span>
      Congrats on your first Tilt resource 🎉 Opt into analytics? (Read more
       
      <a
        href="https://github.com/windmilleng/tilt#privacy"
        target="_blank"
      >
        here
      </a>
      .) 
    </span>
    <span
      className="AnalyticsNudge-buttons"
    >
      <button
        onClick={[Function]}
      >
        Yes!
      </button>
      <button
        onClick={[Function]}
      >
        Nope!
      </button>
    </span>
  </div>
</div>
`;

exports[`shows request-in-progress message 1`] = `
<div
  className="AnalyticsNudge is-visible"
>
  <span>
    Okay, we'll inform the robots...
  </span>
</div>
`;

exports[`shows success message 1`] = `
<div
  className="AnalyticsNudge is-visible"
>
  <span>
    Cool, got it! 👍
  </span>
</div>
`;
